### ISSUE TITLE
2.2.6 Products with global Tier Prices cannot be updated

### ISSUE THREAD

**Issue by davidverholen** [magento/magento2#18154]
### Preconditions

1. Magento 2.2.6

### Steps to reproduce

1. Set Price Scope to Website (Stores -> Configuration -> Catalog -> Catalog -> Prices)
2. Create a Product that has at least one tier price with global scope (all websites)
3. Save Product
4. Try to update Product (reproduced via Rest API - without even sending new tier prices - and Backend Save)

### Expected result

1. Product Values are updated

### Actual result

1. Exception: Unique constraint violation

**The Problem seems to be here:**
https://github.com/magento/magento2/blob/2.2.6/app/code/Magento/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandler.php#L86

While the Interface defines int as return type (not strict, only per type hint), the actual model behind the interface also allows string and returns a string.

As a result $isGlobal is false (although it should be true) because of strict comparison.

This leads to the $old array being empty and finally to Magento trying to insert the values although they already exist in the db (which then causes the Unique Constraint Violation error).

An elegant **workaround** with a plugin on getWebsiteId (cast value to int) seems very risky if in some other place the system expects it to be a string.

**Comment by magento-engcom-team** [comment]
Hi @davidverholen. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@davidverholen do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by davidverholen** [comment]
@magento-engcom-team give me 2.2.6 instance

**Comment by magento-engcom-team** [comment]
Hi @davidverholen. Thank you for your request. I'm working on Magento 2.2.6 instance for you

**Comment by magento-engcom-team** [comment]
Hi @davidverholen, here is your [Magento instance](https://i-18154-2-2-6.engcom.dev.magento.com).
Admin access: [https://i-18154-2-2-6.engcom.dev.magento.com/admin](https://i-18154-2-2-6.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `G1 Passed` will be added to the issue automatically. Please, edit issue description if needed, until label `G1 Passed` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `G2 Passed` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `acknowledged` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report is acknowledged.

**Comment by davidverholen** [comment]
@engcom-backlog-nazar could not reproduce it on vanilla instance so far. investigating further

**Comment by davidverholen** [comment]
ok there is another precondition, added it to the issue:
the price scope in the store configuration has to be set to website

**Comment by davidverholen** [comment]
changed the config on https://i-18154-2-2-6.engcom.dev.magento.com/ and now could reproduce it there

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-95127` were created

**Comment by Sharpy310** [comment]
Just to add some further comments on this issue, we are getting the same issue on our end and we've noticed that if we have tier prices set for customer groups and specific websites on our multisite environment, then when the product is saved as a whole, it's creating an automatic entry on the Advanced Pricing for "All Websites" on the "General" customer group, even though we're not entering that information in, if this is deleted on the Advanced Pricing and re-saved it does appear to work, however it will re-generate the entry, so every time an edit is made on a product you need to delete this entry otherwise you get the Unique Constraint Violation error. See attached screenshot of the record in question.
![4d08589f03f71b23d8bf7621037](https://user-images.githubusercontent.com/17251496/45954435-edfd4200-c004-11e8-93cd-94f9b46f5f91.jpg)

**Comment by FerihumF** [comment]
@davidverholen Thank you for the hint with the plugin on getWebsiteId, so far this is working for me:

di.xml

    <type name="\Magento\Store\Model\Store">
        <plugin name="fix_tier_price_global_error" type="Vendor\Module\Plugin\FixTierPrice" sortOrder="1"/>
    </type>

FixTierPrice.php

    <?php
    namespace Vendor\Module\Plugin;

    class FixTierPrice {

        public function afterGetWebsiteId(\Magento\Store\Model\Store $subject, $result) {
            if (is_string($result)) {
                return intval($result);
            }
            return $result;
        }
    }

I let you know, if I experience any problems. But since getWebsiteId allows int as type I don´t think there should be any issues.

**Comment by reijervano** [comment]
I am afraid the plugin is not enough to fix all the problems with tierpricing. Now it is not possible to create multiple tierprices for a specific website. You always get an Unique constraint violation. This is because of the following check in the vendor/magento/module-catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandler.php class:

```
private function prepareOriginalDataToCompare($origPrices, $isGlobal = true): array
    {
        $old = [];
        if (is_array($origPrices)) {
            foreach ($origPrices as $data) {
                if ($isGlobal === $this->isWebsiteGlobal((int)$data['website_id'])) {
                    $key = $this->getPriceKey($data);
                    $old[$key] = $data;
                }
            }
        }

        return $old;
    }
```
It will causes the "old" array to be emtpy when you try to save or update website specific tierprices. We fixed this by overwriting the specific class.

Can anyone explain me why the check should be there?

**Comment by Detzler** [comment]
I've updated SaveHandler.php and UpdateHandler.php as described in the previous comments. It's now possible to save the tier/group-prices at the global scope. When trying to save at the website scope, the unique constraint violation exception appears.

Does anybody have a workaround/fix? (I take also the dirty one)

**Comment by Detzler** [comment]
I've backported SaveHandler.php and UpdateHandler.php from 2.3-develop into 2.2.6
Now I can save also at the website scope.

**Comment by agorbulin** [comment]
@magento-engcom-team give me 2.2.7 instance

**Comment by magento-engcom-team** [comment]
Hi @agorbulin. Thank you for your request. I'm working on Magento 2.2.7 instance for you

**Comment by magento-engcom-team** [comment]
Hi @agorbulin, here is your [Magento instance](https://i-18154-2-2-7.instances.magento-community.engineering).
Admin access: [https://i-18154-2-2-7.instances.magento-community.engineering/admin](https://i-18154-2-2-7.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by speedupmate** [comment]
any insight on this issue , especially on part @reijervano pointed out , why is this condition there and what does it affect. Pretty huge issue when your site can't save items

**Comment by PascalBrouwers** [comment]
@Detzler can you make a pull request for that backport for 2.2?

**Comment by speedupmate** [comment]
@hostep searched this out for us and  here's a commit that solves this https://github.com/magento/magento2/commit/9240974c8c7

**Comment by erikhansen** [comment]
@Detzler Thanks for your tip on using the [`SaveHandler.php`](https://raw.githubusercontent.com/magento/magento2/2.3-develop/app/code/Magento/Catalog/Model/Product/Attribute/Backend/TierPrice/SaveHandler.php) and [`UpdateHandler.php`](https://raw.githubusercontent.com/magento/magento2/2.3-develop/app/code/Magento/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandler.php) from 2.3-develop. I converted the difference between those two files on 2.2.7 vs 2.3-develop and applied it to a Magento 2.2.7 install using [composer-patches](https://github.com/cweagans/composer-patches).

Here's the patch in case anyone's interested: 
[github-issue-18154.patch.zip](https://github.com/magento/magento2/files/2699486/github-issue-18154.patch.zip)

**Comment by ihor-sviziev** [comment]
Hi @davidverholen. Thank you for your report. 
The issue has been fixed in 2.2-develop branch
Related commit(s):
 - [1c7b6ccd4db558539da93509828230fd4992fbdf](https://github.com/magento/magento2/commit/1c7b6ccd4db558539da93509828230fd4992fbdf)

The fix will be available with the upcoming 2.2.8 release.

**Comment by maxxeh1** [comment]
@erikhansen does this patch file work for 2.2.7?
